components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/PingScheduler.kt

```diff
@@ -0,0 +1,213 @@
+    val pingStorageEngine = pingStorageEngine
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val metricsPingScheduler = metricsPingScheduler
+
+    companion object {
+        /**
+         * Helper function to check to see if a [PingUploadWorker] has been scheduled by the
+         * [PingScheduler]
+         *
+         * @return True if [PingUploadWorker] task found in [WorkManager], false otherwise
+         */
+        internal fun isWorkScheduled(): Boolean {
+            val instance = WorkManager.getInstance()
+            val statuses = instance.getWorkInfosByTag(PingUploadWorker.PING_WORKER_TAG)
+            try {
+                val workInfoList = statuses.get()
+                for (workInfo in workInfoList) {
+                    val state = workInfo.state
+                    if ((state === WorkInfo.State.RUNNING) || (state === WorkInfo.State.ENQUEUED)) {
+                        return true
+                    }
+                }
+            } catch (e: ExecutionException) {
+                // Do nothing but will return false
+            } catch (e: InterruptedException) {
+                // Do nothing but will return false
+            }
+
+            return false
+        }
+    }
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    /**
+     * Handles necessary glean tasks upon entering the background state [Lifecycle.Event.ON_STOP]
+     */
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun onEnterBackground() {
+        // We're going to background, so store how much time we spent
+        // on foreground.
+        GleanBaseline.duration.stopAndSum()
+
+        if (!Glean.isInitialized()) {
+            logger.error("Glean not initialized, abort serialize and send ping")
+            return
+        }
+
+        // Do not send ping if upload is disabled
+        if (!Glean.getUploadEnabled()) {
+            logger.debug("Glean upload not enabled, abort serialize and send ping.")
+            return
+        }
+
+        // Schedule the baseline and event pings and save the result to determine whether or not
+        // we need to schedule the upload worker
+        var availableToSend = assembleAndSerializePing(BaselinePing.STORE_NAME)
+        availableToSend = availableToSend || assembleAndSerializePing("events")
+
+        // Check the metrics ping schedule to determine whether it's time to schedule a new
+        // metrics ping or not
+        if (metricsPingScheduler.canSendPing()) {
+            availableToSend = availableToSend || assembleAndSerializePing(MetricsPingScheduler.STORE_NAME)
+            metricsPingScheduler.updateSentTimestamp()
+        }
+
+        // It should only take a single PingUploadWorker to process all queued pings, so we only
+        // want to have one scheduled at a time.
+        if (availableToSend && !isWorkScheduled()) {
+            WorkManager.getInstance().enqueue(buildWorkRequest())
```

Comment:
With the `isWorkScheduled` check it looks like you've rebuild [enqueueUniqueWork](https://developer.android.com/reference/androidx/work/WorkManager.html#enqueue(androidx.work.WorkRequest))?

Reply:
Sweet!  Yes it does, and that allows me to move the `isWorkScheduled` to be a test only function in TestUtil.kt   Much better!